app/controllers/auth/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class Auth::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def self.provides_callback_for(provider)
+    provider_id = provider.to_s.chomp '_oauth2'
+    define_method provider do
+      @user = User.find_for_oauth(env['omniauth.auth'], current_user)
+
+      if @user.persisted?
+        sign_in_and_redirect @user, event: :authentication
+        set_flash_message(:notice, :success, kind: provider_id.capitalize) if is_navigational_format?
+      else
+        session["devise.#{provider}_data"] = env['omniauth.auth']
+        redirect_to new_user_registration_url
+      end
+    end
+  end
+
+  [:ldap, :cas].each do |provider|
```

Comment:
As said in https://github.com/tootsuite/mastodon/pull/4032#issuecomment-312914005 LDAP doesn't suit in this process